include/ensmallen_bits/callbacks/report.hpp

```diff
@@ -0,0 +1,558 @@
+  void Evaluate(OptimizerType& /* optimizer */,
+                FunctionType& /* function */,
+                const MatType& /* coordinates */,
+                const double objectiveIn)
+  {
+    objective = objectiveIn;
+    evaluateCalls++;
+  }
+
+  /**
+    * Callback function called at any call to EvaluateConstraint().
+    *
+    * @param optimizer The optimizer used to update the function.
+    * @param function Function to optimize.
+    * @param coordinates Starting point.
+    * @param constraint The index of the constraint;
+    * @param objective Objective value of the current point.
+    */
+  template<typename OptimizerType, typename FunctionType, typename MatType>
+  void EvaluateConstraint(OptimizerType& optimizer,
+                          FunctionType& function,
+                          const MatType& coordinates,
+                          const size_t /* constraint */,
+                          const double objective)
+  {
+    Evaluate(optimizer, function, coordinates, objective);
+  }
+
+  /**
+   * Callback function called at any call to Gradient().
+   *
+   * @param optimizer The optimizer used to update the function.
+   * @param function Function to optimize.
+   * @param coordinates Starting point.
+   * @param gradientIn Matrix that holds the gradient.
+   */
+  template<typename OptimizerType, typename FunctionType, typename MatType>
+  void Gradient(OptimizerType& /* optimizer */,
+                FunctionType& /* function */,
+                const MatType& /* coordinates */,
+                const MatType& gradientIn)
+  {
+    hasGradient = true;
+    gradientNorm = arma::norm(gradientIn);
+    gradientCalls++;
+  }
+
+  /**
+   * Callback function called at any call to GradientConstraint().
+   *
+   * @param optimizer The optimizer used to update the function.
+   * @param function Function to optimize.
+   * @param coordinates Starting point.
+   * @param constraint The index of the constraint;
+   * @param gradient Matrix that holds the gradient;
+   */
+  template<typename OptimizerType, typename FunctionType, typename MatType>
+  void GradientConstraint(OptimizerType& optimizer,
+                          FunctionType& function,
+                          const MatType& coordinates,
+                          const size_t /* constraint */,
+                          const MatType& gradient)
+  {
+    Gradient(optimizer, function, coordinates, gradient);
+  }
+
+ private:
+  /**
+   * Helper function to print the number of function to the specified output
+   * stream.
+   *
+   * @param function The instantiated function that implements NumFunctions().
+   * @param stream The output stream.
+   */
+  template<typename FunctionType>
+  typename std::enable_if<
+      traits::HasNumFunctionsSignature<FunctionType>::value, void>::type
+  PrintNumFunctions(const FunctionType& function, std::stringstream& stream)
+  {
+    PrettyPrintElement(stream, "Num functions:", 30);
```

Comment:
```suggestion
    PrettyPrintElement(stream, "Number of functions:", 30);
```
I dunno, maybe this makes it too verbose, I'm not sure.  "Number of objectives"?  I'm struggling with the best terminology here.